From 615c9f71840dedb58bb1c391c988cd6accbf2642 Mon Sep 17 00:00:00 2001
From: Bart Ribbers <bribbers@disroot.org>
Date: Mon, 3 Feb 2025 12:07:03 +0100
Subject: [PATCH] Use the XSI version of strerror_r when the GNU version isn't
 available

Required for Musl systems and thus Alpine Linux and postmarketOS
---
 flatpak-integrator/plugin.cpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/flatpak-integrator/plugin.cpp b/flatpak-integrator/plugin.cpp
index b29b4482..27f064cc 100644
--- a/flatpak-integrator/plugin.cpp
+++ b/flatpak-integrator/plugin.cpp
@@ -143,7 +143,14 @@ private:
     {
         constexpr auto maxBufferSize = 1024;
         thread_local std::array<char, maxBufferSize> buffer;
+#ifdef STRERROR_R_CHAR_P
         return strerror_r(error, buffer.data(), buffer.size());
+#else
+        // Won't be changed by strerror_r but not const so compiler doesn't throw an error
+        static char unknown[] = "unknown error";
+
+        return strerror_r(error, buffer.data(), buffer.size()) ? unknown : buffer.data();
+#endif
     }
 
     int openNoSymlinks(const std::filesystem::path &path, int flags, mode_t mode = 0)
-- 
GitLab

